src/charmcraft/charmcraft.js

```diff
@@ -0,0 +1,109 @@
+const core = require("@actions/core");
+const exec = require("@actions/exec");
+const fs = require("fs");
+const yaml = require("js-yaml");
+const glob = require("@actions/glob");
+
+/* eslint-disable camelcase */
+
+class Charmcraft {
+  constructor() {
+    this.uploadImage = core.getInput("upload-image").toLowerCase() === "true";
+    core.exportVariable("CHARMCRAFT_AUTH", core.getInput("credentials"));
+  }
+
+  async uploadResources() {
+    let resourceInfo = "resources:\n";
+
+    const { name, images } = this.metadata();
+    const flags = await Promise.all(
+      images.map(async ([resource_name, resource_image]) => {
+        await this.uploadResource(resource_image, name, resource_name);
+        const { flag, info } = await this.buildResourceFlag(
+          name,
+          resource_name,
+          resource_image
+        );
+        resourceInfo += info;
+        return flag;
+      })
+    );
+    return { flags, resourceInfo };
+  }
+
+  async uploadResource(resource_image, name, resource_name) {
+    if (!this.uploadImage) {
+      core.warning(
+        "No resources where uploaded as part of this build.\n" +
+          `If you wish to upload the OCI image, set 'upload-image' to 'true'`
+      );
+      return;
+    }
+
+    await exec.exec("docker", ["pull", resource_image]);
+    await exec.exec("charmcraft", [
+      "upload-resource",
+      "--quiet",
+      name,
+      resource_name,
+      "--image",
+      resource_image,
+    ]);
+  }
+
+  async buildResourceFlag(name, resource_name, resource_image) {
+    const result = await exec.getExecOutput("charmcraft", [
+      "resource-revisions",
+      name,
+      resource_name,
+    ]);
+
+    const revision = result.stdout.split("\n")[1].split(" ")[0];
```

Comment:
Probably worth a comment here to show what you're parsing for when it inevitably changes!

@facundobatista having the option for Charmcraft to have a more global JSON (or similar) output mode would be nice

Reply:
Yeah, an `--out json` would definitely make life easier, and less brittle 😅